packages/api/internal/handlers/sandboxes_list.go

```diff
@@ -152,6 +105,118 @@ func (a *APIStore) GetSandboxes(c *gin.Context, params api.GetSandboxesParams) {
 		sandboxes = append(sandboxes, instance)
 	}
 
+	// append latest snapshots to sandboxes
+	for _, e := range snapshotEnvs {
+		snapshotBuilds := e.Edges.Builds
+		if len(snapshotBuilds) == 0 {
+			continue
+		}
+
+		snapshot := e.Edges.Snapshots[0]
+
+		instance := api.RunningSandbox{
+			ClientID:   "00000000",
+			TemplateID: e.ID,
+			SandboxID:  snapshot.SandboxID,
+			StartedAt:  snapshot.SandboxStartedAt,
+			CpuCount:   int32(snapshotBuilds[0].Vcpu),
+			MemoryMB:   int32(snapshotBuilds[0].RAMMB),
+			EndAt:      snapshot.PausedAt,
+			State:      "paused",
+		}
+
+		if snapshot.Metadata != nil {
+			meta := api.SandboxMetadata(snapshot.Metadata)
+			instance.Metadata = &meta
+		}
+
+		sandboxes = append(sandboxes, instance)
+	}
+
+	// filter sandboxes by metadata
+	if params.Query != nil {
+		// Unescape query
+		query, err := url.QueryUnescape(*params.Query)
+		if err != nil {
+			c.JSON(http.StatusBadRequest, "Error when unescaping query")
+
+			return
+		}
+
+		// Parse filters, both key and value are also unescaped
+		filters := make(map[string]string)
+
+		for _, filter := range strings.Split(query, "&") {
+			parts := strings.Split(filter, "=")
+			if len(parts) != 2 {
+				c.JSON(http.StatusBadRequest, "Invalid key value pair in query")
+
+				return
+			}
+
+			key, err := url.QueryUnescape(parts[0])
+			if err != nil {
+				c.JSON(http.StatusBadRequest, "Error when unescaping key")
+
+				return
+			}
+
+			value, err := url.QueryUnescape(parts[1])
+			if err != nil {
+				c.JSON(http.StatusBadRequest, "Error when unescaping value")
+
+				return
+			}
+
+			filters[key] = value
+		}
+
+		// Filter instances to match all filters
+		n := 0
+		for _, instance := range sandboxes {
+			if instance.Metadata == nil {
+				continue
+			}
+
+			matchesAll := true
+			for key, value := range filters {
+				if metadataValue, ok := (*instance.Metadata)[key]; !ok || metadataValue != value {
+					matchesAll = false
+					break
+				}
+			}
+
+			if matchesAll {
+				sandboxes[n] = instance
+				n++
+			}
+		}
+
+		// Trim slice
+		sandboxes = sandboxes[:n]
+	}
+
+	// filter sandboxes by state
+	if params.State != nil {
```

Comment:
I think you should handle it differently, check the state at start and do only the work that's needed, listing all snapshots if you don't need them can be very wasteful